Fix BaseOutput initialization from dict

anton-l opened the pull request:
This will allow `accelerate` to reinitialize model outputs like so: `output = UNet2DOutput({"sample": torch.randn(3, 4)})` and avoid nesting attributes.

Closes https://github.com/huggingface/diffusers/issues/560

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_570). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented:
Now I see the need for this code :sweat_smile: . We need to adapt it though for `diffusers`. 
I've deleted it here: https://github.com/huggingface/diffusers/pull/422/files (sorry should have explained better in the PR why)

But essentially, the code as is on this branch will currently break the following:

```python
from diffusers import DDIMPipeline

pipeline = DDIMPipeline.from_pretrained("google/ddpm-cifar10-32")
outputs = pipeline()

outputs["images"]
```

@anton-l could you see how this can be fixed and then maybe also add a test here :-)

anton-l commented on tests/test_outputs.py:
```diff
@@ -0,0 +1,60 @@
+        # check every way of getting the attribute
+        assert isinstance(outputs.images, np.ndarray)
+        assert outputs.images.shape == (1, 3, 4, 4)
+        assert isinstance(outputs["images"], np.ndarray)
```
The transformers version was failing here

anton-l commented on tests/test_outputs.py:
```diff
@@ -0,0 +1,60 @@
+        outputs = CustomOutput({"images": np.random.rand(1, 3, 4, 4)})
+
+        # check every way of getting the attribute
+        assert isinstance(outputs.images, np.ndarray)
```
the pre-PR diffusers version was failing here

anton-l commented:
@patrickvonplaten I've implemented a simpler version that works for your case as well.
Also added high-coverage tests for every combination of inputs and outputs.

patrickvonplaten commented on tests/test_outputs.py:
```diff
@@ -0,0 +1,60 @@
+
+
+class ConfigTester(unittest.TestCase):
+    def test_outputs_single_attribute(self):
```
Very nice!

patrickvonplaten commented on tests/test_outputs.py:
```diff
@@ -0,0 +1,60 @@
+        outputs = CustomOutput(images=[PIL.Image.new("RGB", (4, 4))])
+
+        # check every way of getting the attribute
+        assert isinstance(outputs.images, list)
```
Cool that this is all tested now!
